sw/device/lib/dif/dif_aes.c

```diff
@@ -0,0 +1,424 @@
+                               AES_STATUS_STALL_BIT);
+}
+
+static bool aes_output_valid(const dif_aes_t *aes) {
+  return mmio_region_get_bit32(aes->params.base_addr, AES_STATUS_REG_OFFSET,
+                               AES_STATUS_OUTPUT_VALID_BIT);
+}
+
+static bool aes_input_ready(const dif_aes_t *aes) {
+  return mmio_region_get_bit32(aes->params.base_addr, AES_STATUS_REG_OFFSET,
+                               AES_STATUS_INPUT_READY_BIT);
+}
+
+static bool aes_alert_fatal(const dif_aes_t *aes) {
+  return mmio_region_get_bit32(aes->params.base_addr, AES_STATUS_REG_OFFSET,
+                               AES_STATUS_ALERT_FATAL_BIT);
+}
+
+static bool aes_alert_recoverable(const dif_aes_t *aes) {
+  return mmio_region_get_bit32(aes->params.base_addr, AES_STATUS_REG_OFFSET,
+                               AES_STATUS_ALERT_RECOVERABLE_BIT);
+}
+
+static void aes_shadowed_write(mmio_region_t base, ptrdiff_t offset,
+                               uint32_t value) {
+  mmio_region_write32(base, offset, value);
+  mmio_region_write32(base, offset, value);
+}
+
+static void aes_clear_internal_state(const dif_aes_t *aes) {
+  // Make sure AES is idle before clearing.
+  while (!aes_idle(aes)) {
+  };
+
+  // It should be fine to clobber the Control register. Only
+  // `AES_CTRL_SHADOWED_MANUAL_OPERATION` bit must be set.
+  uint32_t ctrl_reg =
+      bitfield_bit32_write(0, AES_CTRL_SHADOWED_MANUAL_OPERATION_BIT, true);
+
+  aes_shadowed_write(aes->params.base_addr, AES_CTRL_SHADOWED_REG_OFFSET,
+                     ctrl_reg);
+
+  uint32_t trigger_reg =
+      bitfield_bit32_write(0, AES_TRIGGER_KEY_IV_DATA_IN_CLEAR_BIT, true);
+  trigger_reg =
+      bitfield_bit32_write(trigger_reg, AES_TRIGGER_DATA_OUT_CLEAR_BIT, true);
+
+  mmio_region_write32(aes->params.base_addr, AES_TRIGGER_REG_OFFSET,
+                      trigger_reg);
+
+  // Make sure AES is cleared before proceeding (may take multiple cycles).
+  while (!aes_idle(aes)) {
+  };
+}
+
+static aes_key_field_val_t key_to_field(dif_aes_key_length_t key) {
+  switch (key) {
+    case kDifAesKey128:
+      return kAesKeyFieldVal128;
+    case kDifAesKey192:
+      return kAesKeyFieldVal192;
+    case kDifAesKey256:
+      return kAesKeyFieldVal256;
+    default:
+      return kAesKeyFieldValInvalid;
+  }
+}
+
+/**
+ * Configures AES. Is used by every `dif_aes_start_<mode>` function.
+ *
+ * @param aes AES state data.
+ * @param transaction Configuration data, common across all Cipher modes.
+ * @param cipher_mode_val Cipher Mode register write value.
+ * @return `dif_aes_start_result_t`.
+ */
+static dif_aes_start_result_t configure(
+    const dif_aes_t *aes, const dif_aes_transaction_t *transaction,
+    aes_mode_field_val_t cipher_mode_val) {
+  aes_key_field_val_t key_len_val = key_to_field(transaction->key_len);
+  if (key_len_val == kAesKeyFieldValInvalid) {
+    return kDifAesStartError;
+  }
+
+  // TODO - it is probably fine to clobber the CTRL register in process?
```

Comment:
File a bug...?

Reply:
> File a bug...?

You mean an issue?

Anyhow - this has been removed.